File: src/extras-kfilemetadata/libwritercollection.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/kfilemetadata-writercollection.html)
pub const kfilemetadata__writercollection = struct {
    /// New constructs a new KFileMetaData::WriterCollection object.
    ///
    pub fn New() QtC.KFileMetaData__WriterCollection {
        return qtc.KFileMetaData__WriterCollection_new();
    }

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-writercollection.html#fetchWriters)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QtC.KFileMetaData__WriterCollection `
    ///
    /// ` mimetype: []const u8 `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn FetchWriters(self: ?*anyopaque, mimetype: []const u8, allocator: std.mem.Allocator) []QtC.KFileMetaData__Writer {
        const mimetype_str = qtc.libqt_string{
            .len = mimetype.len,
            .data = mimetype.ptr,
        };
        const _arr: qtc.libqt_list = qtc.KFileMetaData__WriterCollection_FetchWriters(@ptrCast(self), mimetype_str);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.KFileMetaData__Writer, _arr.len) catch @panic("kfilemetadata__writercollection.FetchWriters: Memory allocation failed");
        const _data: [*]QtC.KFileMetaData__Writer = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ## Parameter:
    ///
    /// ` self: QtC.KFileMetaData__WriterCollection `
    ///
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KFileMetaData__WriterCollection_Delete(@ptrCast(self));
    }
};

File: src/extras-kfilemetadata/libwritercollection.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_KFileMetaData__Writer
#define WORKAROUND_INNER_CLASS_DEFINITION_KFileMetaData__WriterCollection
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <writercollection.h>
#include "libwritercollection.h"
#include "libwritercollection.hxx"

KFileMetaData__WriterCollection* KFileMetaData__WriterCollection_new() {
    return new KFileMetaData::WriterCollection();
}

libqt_list /* of KFileMetaData__Writer* */ KFileMetaData__WriterCollection_FetchWriters(const KFileMetaData__WriterCollection* self, const libqt_string mimetype) {
    QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
    QList<KFileMetaData::Writer*> _ret = self->fetchWriters(mimetype_QString);
    // Convert QList<> from C++ memory to manually-managed C memory
    KFileMetaData__Writer** _arr = static_cast<KFileMetaData__Writer**>(malloc(sizeof(KFileMetaData__Writer*) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = _ret[i];
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void KFileMetaData__WriterCollection_Delete(KFileMetaData__WriterCollection* self) {
    delete self;
}
